rialto_airflow/harvest/pubmed.py

```diff
@@ -0,0 +1,196 @@
+
+                    with get_session(snapshot.database_name).begin() as insert_session:
+                        # if there's a DOI constraint violation we need to update instead of insert
+                        pub_id = insert_session.execute(
+                            insert(Publication)
+                            .values(
+                                doi=doi,
+                                pubmed_json=pubmed_pub,
+                            )
+                            .on_conflict_do_update(
+                                constraint="publication_doi_key",
+                                set_=dict(pubmed_json=pubmed_pub),
+                            )
+                            .returning(Publication.id)
+                        ).scalar_one()
+
+                        # a constraint violation is ok here, since it means we
+                        # already know that the publication is by the author
+                        insert_session.execute(
+                            insert(pub_author_association)
+                            .values(publication_id=pub_id, author_id=author.id)
+                            .on_conflict_do_nothing()
+                        )
+
+                        jsonl_output.write(json.dumps(pubmed_pub) + "\n")
+
+    return jsonl_file
+
+
+def pmids_from_orcid(orcid):
+    """
+    Returns PMIDs associated with a given ORCID.
+    """
+    # Pubmed doesn't want the full ORCID URIs which are stored in User table
+    if m := re.match(r"^https?://orcid.org/(.+)$", orcid):
+        orcid = m.group(1)
+
+    return _pubmed_search_api(f"{orcid}[auid]")
+
+
+def publications_from_pmids(pmids):
+    """
+    Returns full pubmed records given a list of PMIDs.
+    """
+    if not pmids or not isinstance(pmids, list):
+        return None
+
+    query = "id=" + "&id=".join(pmids)
+
+    http = requests.Session()
+
+    full_url = (
+        f"{BASE_URL}{FETCH_PATH}&api_key={os.environ.get('AIRFLOW_VAR_PUBMED_KEY')}"
+    )
+    logging.info(f"fetching full records from pubmed with {query}")
+    resp: requests.Response = http.post(full_url, params=query, headers=HEADERS)
+
+    results = resp.content
+
+    json_results = xmltodict.parse(results)
+    pubs = json_results["PubmedArticleSet"]["PubmedArticle"]
+    if not isinstance(pubs, list):
+        # if there is only one record, it will not be in a list, but we want to be in one so we can iterate over it
+        return [pubs]
+    else:
+        return pubs
+
+
+def _pubmed_search_api(query) -> list:
+    """
+    Return a list of pmids given a general search query.
+    """
+
+    params: Params = {"term": query}
+
+    http = requests.Session()
+
+    full_url = (
+        f"{BASE_URL}{SEARCH_PATH}&api_key={os.environ.get('AIRFLOW_VAR_PUBMED_KEY')}"
+    )
+    logging.info(f"searching pubmed with {params}")
+    resp: requests.Response = http.get(full_url, params=params, headers=HEADERS)
```

Comment:
it might be worth calling `resp.raise_for_status()` here which will make sure an exception is thrown unless we get back a HTTP Status that went ok: 

https://3.python-requests.org/api/#requests.Response.raise_for_status

Reply:
will do this